storage-proofs/src/merkle.rs

```diff
@@ -153,6 +165,168 @@ fn path_index<T: Domain>(path: &[(T, bool)]) -> usize {
+// to set the file name (to label trees with the number of the corresponding
+// layer they belong to) and `DISK_BACKED_MMAP_TREE_DIR` is set externally by
+// the user to define the directory where the files will be created. (If one
+// of these is not set we just create a random file that will be removed after
+// `DiskMmapStore` is dropped).
+//
+// FIXME: Copied from `MmapStore`, needs to be integrated into
+//  `merkle_light::merkle`. The only difference is in the `Store::new()`
+//  method which creates a file mapping (instead of an anonymous one).
+//
+// FIXME: `DISK_BACKED_MMAP_TREE_DIR` is being manually set in local
+//  tests.
+#[derive(Debug)]
+pub struct DiskMmapStore<E: Element> {
+    store: MmapMut,
+    len: usize,
+    _e: PhantomData<E>,
+    file: File,
+    // We need to save the `File` in case we're creating a `tempfile()`
+    // otherwise it will get cleaned after we return from `new()`.
+    // FIXME: Check this (at least that was the case for `tempdir()`).
+}
+
+impl<E: Element> ops::Deref for DiskMmapStore<E> {
+    type Target = [E];
+
+    fn deref(&self) -> &Self::Target {
+        unimplemented!()
+    }
+}
+
+impl<E: Element> Store<E> for DiskMmapStore<E> {
+    #[allow(unsafe_code)]
+    fn new(size: usize) -> Self {
+        let byte_len = E::byte_len() * size;
+
+        let file: File = if env::var("INTERNAL_TREE_NAME").is_ok()
+            && env::var("DISK_BACKED_MMAP_TREE_DIR").is_ok()
+        {
+            let trees_dir = PathBuf::from(env::var("DISK_BACKED_MMAP_TREE_DIR").unwrap());
+
+            // Try to create `trees_dir`, ignore the error if `AlreadyExists`.
+            if let Some(create_error) = fs::create_dir(&trees_dir).err() {
+                if create_error.kind() != io::ErrorKind::AlreadyExists {
+                    panic!(create_error);
+                }
+            }
+
+            let tree_name = env::var("INTERNAL_TREE_NAME").unwrap();
+
+            let tree_name = format!("{}-{}", tree_name, rand::random::<u32>());
+            // FIXME: The user of `DISK_BACKED_MMAP_TREE_DIR` should figure out
+            // how to manage this directory, for now we create every file with
+            // a different random number, the problem being that tests now do
+            // replications many times in the same run so they end up reusing
+            // the same files with invalid (old) data.
+
+            let tree_path = &trees_dir.join(&tree_name);
+
+            info!(SP_LOG, "creating tree mmap-file"; "path" => &tree_path.to_str());
+
+            OpenOptions::new()
+                .read(true)
+                .write(true)
+                .create(true)
+                .open(&tree_path)
+                .unwrap()
+        } else {
+            // We don't know which layer this is (`INTERNAL_TREE_NAME`) or we
+            // don't know where to save it (`DISK_BACKED_MMAP_TREE_DIR`), just
+            // create *any* temporary file.
+            tempfile().unwrap()
+        };
```

Comment:
Please see the comment above, https://github.com/filecoin-project/rust-fil-proofs/pull/567#issuecomment-478653789, if you could orient me towards a concrete API implementation that would be very helpful.

Reply:
Let's talk this through tomorrow if what I say here doesn't make sense, but something like:

- For `DiskMmapStore`, I would do exactly as described above (i.e. extend `new` to take an optional path.
- For `MerkleTree`, I would do as @dignifiedquire suggests:
  - Add a method which returns the required size for a given `data` input.
- Add a new type `DiskMerkleTree` which uses a `DiskMmapStore` (analogous to `MerkleTree` using `MmapStore`).
- Add a bolean config var, say, `REPLICATE-WITH-DISK-TREES` (or some better name).
- Use either `MerkleTree` or `DiskMerkleTree` when replicating, depending on the value of that config option. (NOTE: don't check for env vars directly in the code, use `config::get_config`.

Reply:
@porcuquine 

> * For `DiskMmapStore`, I would do exactly as described above (i.e. extend `new` to take an optional path.
> * For `MerkleTree`, I would do as @dignifiedquire suggests:
>   
>   * Add a method which returns the required size for a given `data` input.

As mentioned above I still don't see how this fits with the existing `Store` trait (which doesn't have a file concept and should be extended with one, which would panic with the current implementations laying the burden on the user to be careful not to use it with the normal `MerkleTree`; I'm not saying this is a bad option just that a comment addressing this would be useful to me). I'm definitely missing something of what both you and @dignifiedquire are trying to explain me.

Reply:
It's entirely possible that I'm the one missing it, since I'm only observing this work from a slight remove.

That said, here's what I *think*:
 - `Store` is very general and doesn't even know about the distinction between a `Vec` and `Mmap` i,mplementation.
 - `MmapStore` doesn't (and doesn't need to) know about files (but also, therefore, doesn't accomplish the goals of our disk-backed merkle trees).

It's not obvious to me why `Store` would need to become aware of files. That some particular `Store` implementation is backed by concrete files is an implementation detail that *shouldn't* be leaked into the more general trait. If it were, that would make instantiating an ordinary vector-based merkle tree awkward (for example).

I feel like we might be saying the same thing in some ways. It's just not clear to me why you think we need to push the details of a more specific type implementing the `Store` trait up into that trait. Doing that would seem to cause the problems you note, so why would we do it?

It's possible that @dignifiedquire (who has tracked and participated in this work a little more closely) understand the issue better, though.

In case what I'm saying does in fact make sense, the focusing question is: what would go wrong if we proceeded roughly as I described above and didn't modify the `Store` trait at all?

Reply:
What you're saying sounds right, maybe it's my Rust inexperience that led to assume that the only way was to modify the `Store` trait, I'll prepare a PR on the `merkle_light` side with this (even if I can't make it work it will give me a more concrete code to move forward with this discussion). Thanks for looking into this.

Reply:
@porcuquine Quick follow-up question to your proposed solution, how do I switch between `DiskMerkleTree` and (the non-path version) `MerkleTree`? That is, using our dynamic configuration framework how can I select between two statically different types? (Again, I might be missing some Rust related aspect of the solution.)

Reply:
The way to do it is to use a trait object, i.e. `Box<merkle::MerkleTree>`. That said, this *will* probably mean propagating a lot of changes.

Given that we probably don't actually need to eventually support both, it might make morse sense to just define another type `DefaultMerkleTree` (or whatever) which can be set as appropriate to switch at compile time.

We could add a feature flag to switch between the two if we're nervous about starting out with disk-backed as the default (where we probably want to end up). Something like `[cfg(disk-trees)]` and `[cfg(not(disk-trees))]`.

The main reason not to just go with disk trees to begin with — assuming we can verify no performance loss — is the cleanup issue.

The one thing to avoid is spending a lot of time trying one way, failing, then sinking more time into the other. For that reason, and because I also because it's probably closer to where we want to end up, I think I'd probably recommend just going with the compile-time flag instead of the run-time config.